File: Problems/Algorithms/235. Lowest Common Ancestor of a Binary Search Tree/LowestCoommonAncestorBST.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        if ((root!!.`val` - p!!.`val`) * (root!!.`val` - q!!.`val`) <= 0) return root
        
        if (root!!.`val` > q!!.`val` && root!!.`val` > p!!.`val`) {
            return lowestCommonAncestor(root?.left, p, q)
        }
        return lowestCommonAncestor(root?.right, p, q)
    }
}

File: Problems/Algorithms/235. Lowest Common Ancestor of a Binary Search Tree/lowest_coommon_ancestor_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if ((root.val - p.val) * (root.val - q.val) <= 0):
            return root

        if root.val > q.val and root.val > p.val:
            return self.lowestCommonAncestor(root.left, p, q)
        return self.lowestCommonAncestor(root.right, p, q)

